Complete porting the various Blocks for Java

johnoliver opened the pull request:
### Motivation and Context
<!-- Thank you for your contribution to the semantic-kernel repo!
Please help reviewers and future users, providing the following information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->


### Description
<!-- Describe your changes, the overall approach, the underlying design.
     These notes will help understanding how your code works. Thanks! -->


### Contribution Checklist
<!-- Before submitting this PR, please make sure: -->
- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [ ] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/DefaultPromptTemplateEngine.java:
```diff
@@ -76,13 +78,15 @@ public Mono<String> renderAsync(List<Block> blocks, SKContext context) {
+                                        new TemplateException(
+                                                TemplateException.ErrorCodes.UnexpectedBlockType,
+                                                "Unexpected block type, the block doesn't have a"
+                                                        + " rendering method"));
```
Create a variable to contain the message. The string concatenation creates extra strings and it made worse that we have the same string split up different.

johnoliver replied:
the splitting up is an artifact of the code formatter enforcing line lengths and splitting the string to maintain a line length. This causes no issues as the compiler eliminates the concatenation, so with or without the concatenation you will end up with the same bytecode. But can extract it to a variable if you feel makes more readable code, as I say performance wise extracted, with or without concatenation, it all gets compiled to the same bytecode.

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+public class TemplateException extends SKException {
+
+    @Nonnull private final ErrorCodes errorCode;
+
```
```suggestion

/**
 * <p>
 * Initializes a new instance of the {@code TemplateException} class with a provided error code.
 * </p>
 * @param error The error code.
 */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+    public TemplateException(@Nonnull ErrorCodes error) {
+        this(error, null, null);
+    }
+
```
```suggestion

/**
 * <p>
 * Initializes a new instance of the {@code TemplateException} class with a provided error code.
 * </p>
 * @param error The error code.
 * @param message The exception message.
 */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+    public TemplateException(@Nonnull ErrorCodes errorCode, @Nullable String message) {
+        this(errorCode, message, null);
+    }
+
```
```suggestion

/**
 * <p>
 * Initializes a new instance of the {@code TemplateException} class with a provided error code.
 * </p>
 * @param error The error code.
 * @param message The exception message.
 * @param innerException The exception that is the cause of the current exception.
 */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+        super(getDefaultMessage(errorCode, message), innerException);
+        this.errorCode = errorCode;
+    }
+
```
``` suggestion

/**
 * <p>
 * Gets the error code for this exception.
 * </p>
 */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+        return errorCode;
+    }
+
+    /* Translate the error code into a default message */
```
```suggestion
/**
 * <p>
 * Translate the error code into a default message.
 * </p>
 */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+            @Nonnull ErrorCodes errorCode, @Nullable String message) {
+        return String.format("%s: %s", errorCode.getMessage(), message);
+    }
+
```
```suggestion

/**
 * <p>
 * Error codes for {@code TemplateException}.
 * </p>
 */

```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+    }
+
+    public enum ErrorCodes {
+        UnknownError("Unknown error"),
```
```suggestion
      /**
        * <p>
        * Unknown error.
        * </p>
        */
        UnknownError("Unknown error"),
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+
+    public enum ErrorCodes {
+        UnknownError("Unknown error"),
+
```
```suggestion

      /**
        * <p>
        * Syntax error, the template syntax used is not valid.
        * </p>
        */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+        UnknownError("Unknown error"),
+
+        SyntaxError("Syntax error, the template syntax used is not valid"),
+
```
```suggestion

      /**
        * <p>
        * The block type produced be the tokenizer was not expected.
        * </p>
        */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+        SyntaxError("Syntax error, the template syntax used is not valid"),
+
+        UnexpectedBlockType("The block type produced be the tokenizer was not expected"),
+
```
```suggestion

      /**
        * <p>
        * The template requires an unknown function.
        * </p>
        */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/TemplateException.java:
```diff
@@ -3,4 +3,59 @@
+        UnexpectedBlockType("The block type produced be the tokenizer was not expected"),
+
+        FunctionNotFound("The template requires an unknown function"),
+
```
```suggestion


      /**
        * <p>
        * The template execution failed, e.g. a function call threw an exception.
        * </p>
        */
```

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/CodeBlock.java:
```diff
@@ -7,15 +7,21 @@
-// ReSharper disable TemplateIsNotCompileTimeConstantProblem
 public class CodeBlock extends Block implements CodeRendering {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CodeBlock.class);
```
Inject logger rather than hardcode it

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/CodeBlock.java:
```diff
@@ -7,15 +7,21 @@
+import java.util.Optional;
 
-// ReSharper disable TemplateIsNotCompileTimeConstantProblem
 public class CodeBlock extends Block implements CodeRendering {
```
Make this class final

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/TextBlock.java:
```diff
@@ -25,18 +23,4 @@ public String render(ContextVariables variables) {
     public TextBlock(String text, int startIndex, int stopIndex) {
         super(text.substring(startIndex, stopIndex), BlockTypes.Text);
     }
-    /*
```
Why is this implementation not being included?

johnoliver replied:
This class does not offer any functionality over the super class

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/TextBlock.java:
```diff
@@ -6,8 +6,6 @@
 
 public class TextBlock extends Block implements TextRendering {
```
This should be final and package protected

johnoliver replied:
we can make it final, but it is used outside the package

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/ValBlock.java:
```diff
@@ -4,88 +4,63 @@
+
 import reactor.util.annotation.Nullable;
 
 public class ValBlock extends Block implements TextRendering {
```
This should be final and package protected

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/ValBlock.java:
```diff
@@ -4,88 +4,63 @@
 
 public class ValBlock extends Block implements TextRendering {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ValBlock.class);
```
Inject logger

markwallace-microsoft commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/ValBlock.java:
```diff
@@ -4,88 +4,63 @@
+    // Content, excluding start/end quote chars
+    private String value = "";
+
     public ValBlock(String content) {
```
`content` is optional

In c# this variable is named `quotedValue`, any reason to change this?

johnoliver commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/DefaultPromptTemplateEngine.java:
```diff
@@ -25,7 +28,7 @@
 /// </summary>
 public class DefaultPromptTemplateEngine implements PromptTemplateEngine {
-    // private readonly ILogger _log;
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultPromptTemplateEngine.class);
```
Its an unusual pattern for java to inject loggers, general standard practice is to have loggers as statics, is there a usecase that makes injecting loggers important? Due to the lack of default arguments, this will mean you will end up either passing null around a lot when constructing objects, or increasing the number of constructors for all of our objects to add an optional logger.

markwallace-microsoft replied:
I chatted with Shawn about this and no change required at this time. I'm keeping track of where the implementation diverges from .Net and we may discuss this again when we're ready to merge to main.

johnoliver commented on java/semantickernel-core/src/main/java/com/microsoft/semantickernel/templateengine/blocks/VarBlock.java:
```diff
@@ -5,8 +5,12 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 public class VarBlock extends Block implements TextRendering {
```
this gets used by the CodeTokenizer that is in `com.microsoft.semantickernel.templateengine` this is in `com.microsoft.semantickernel.templateengine.blocks`

markwallace-microsoft reviewed (commented):
There are still some suggestions to be committed

johnoliver commented:
> There are still some suggestions to be committed

@markwallace-microsoft With the exception of the logger injection, I believe I have included everything, unfortunately I also accidentally merged in the code from #1131 to this branch including the fixes requested on #1131, I can try to separate the PRs again, however if you are satisfied that I have addressed the issues on #1131 can merge them both here
